CMS: Unify system_name behavior

jlledom opened the pull request:
*What this PR does / why we need it**:

This PR is a suggestion. The `system_name` attribute acts as user-friendly identifier for CMS objects, but it's behavior is inconsistent among different types of objects:

|Type|Required|Format enforced|Value|
|--|--|--|--|
|Section|Yes|No|Generated from `title` parameter|
|Page|No|Yes|From `system_name` parameter|
|Layout|Yes|Yes|From `system_name` parameter|
|Partial|Yes|Yes|From `system_name` parameter|
|File|No|No|Always `nil` (there's no  `system_name`  column in the table) |

I think it'll be simpler to code, test and understand if `system_name` behaves the same way among all CMS items. I created a new `CMS::SystemName` module, being used for sections, pages, layouts and partials; to unify the `system_name` logic.

The new module makes `system_name` always required and format enforced, getting the value from the `system_name` parameter when it exists, and from `title` otherwise.

Files are excluded because they would need a DB migration to support `system_name`.

**Which issue(s) this PR fixes** 

Part of [THREESCALE-8991](https://issues.redhat.com/browse/THREESCALE-8991)

**Verification steps** 

- Creating a new section which title contains spaces should generate a system name without spaces
- Creating a Pages without a system name should generate one automatically, with the proper format

**Special notes for your reviewer**:

Con to consider: updating the validation would require adding a rake task to search and update all invalid system names for existing pages and sections.

mayorova commented:
Nice one!
I haven't reviewed it yet, but a part of it was on my TODO list (I actually thought I opened a PR already, but turns out I didn't).
There was this issue recently: https://issues.redhat.com/browse/THREESCALE-8958
and to prevent this from hapenning again, I was going to add `self.title.downcase` here in [section.rb](https://github.com/3scale/porta/blob/master/app/models/cms/section.rb#L175).

However, this PR seems to incorporate this fix already :+1:

thalesmiguel commented on app/models/cms/section.rb:
```diff
@@ -73,7 +73,7 @@ def partial_paths
 
     def find_or_create!(name, path, options = {})
-      system_name = name.downcase
+      system_name = name.parameterize
```
Why are we changing from `downcase` to `parameterize`?
After this change is deployed, couldn't we end-up with something like
```
Section1 -> my section
Section2 -> my-section
```
because of this `find_or_create!`?

jlledom replied:
`parameterize` is to generate a valid `system_name` for the new format enforcing. In your examples:

```
Section1 -> section1
Section2 -> section2
```

thalesmiguel reviewed (commented):
Nice! Definitely looks like a step in the right direction 🎉

mayorova commented on app/lib/cms/system_name.rb:
```diff
@@ -0,0 +1,20 @@
+  extend ActiveSupport::Concern
+
+  included do
+    validates :system_name, uniqueness: { scope: %i[provider_id type], allow_blank: true },
```
While the new format is the one specified here, I think I would keep the whitespace (only ` `, and not others that belong to `\s`), because I think even in the default template (without customer's custom templates) there would be plenty of system names with whitespaces in them, and it would be a problem in production in SaaS.

We might need to check in the production DB, whether there are any other system names that would not match `\A\w[\w\-/_ ]+\z`. I think most should match.

jlledom replied:
I found that there's already a [SystemName module](https://github.com/3scale/porta/blob/7cdee6ccb1fcc983580bd93972f19101ee3e6c50/app/lib/system_name.rb) which purpose is the same than this one I added. I think the best solution is to addapt the CMS models to include that `SystemName` module and remove mine. But that module it's enforcing the format without a space. I suggest either:

1. Allow whitespaces in that `SystemName` format, so for all the models, not only CMS.
2. Create a rake task to fix all system names in CMS to match the new format. Also update the jobs that create the builtin developer portal for new users.
3. Keep the CMS system names not integrated with other models, so we keep this `app/lib/cms/system_name.rb` to enforce another format with whitespaces only for CMS.

I vote for 2.

mayorova replied:
Hm, I'd prefer 3 + "Also update the jobs that create the builtin developer portal for new users."

My reasoning is that I don't want to touch the existing SaaS data. Even though the API was (and still kind of is) unsupported, we don't know if any of the customers actually rely on the system names for their CMS elements. I would avoid breaking it for them

But I guess we need someone else's opinion to break this tie :wink: 
@akostadinov ?

akostadinov replied:
If system names are used in content to refer to them, then Daria's suggestion makes more sense.

But Joan, if you desire very much consistency, you probably can grep somehow through the portals to check for that, then we can change names. Just be sure that this grepping doesn't kill MySQL or another part of system. And I guess it can take a lot of time. We can discuss this if you want to take on it.

jlledom replied:
It's ok, I'll go with @mayorova's approach

jlledom replied:
After taking a closer look at this, there's no need to update the jobs that create the dev. portal for new users. All builtin content is created sending parameters to the existing models. Since the models now have the new `SystemName ` module included, all `system_names` are generated in the new format. So @mayorova's approach is already implemented.

jlledom commented on app/lib/cms/system_name.rb:
```diff
@@ -0,0 +1,20 @@
+  protected
+
+  def set_system_name
+    self.system_name = title&.parameterize if title && system_name.blank?
```
Changed it to

`self.system_name ||= title.parameterize if title`

 if title can't be removed, otherwise in case of not having title will `system_name` no `nil`

jlledom replied:
Due to a test failing, I had to revert this to:

`self.system_name = title.parameterize if title.present? && system_name.blank?`

`||=` doesn't work properly when `system_name = ""`, so `.blank?` is required

jlledom commented on app/models/cms/email_template.rb:
```diff
@@ -4,7 +4,6 @@ class CMS::EmailTemplate < CMS::Template
 
   attr_accessible :system_name, :draft, :published, :headers # do not do drafts
```
You think correctly

jlledom commented on app/models/cms/email_template.rb:
```diff
@@ -4,7 +4,6 @@ class CMS::EmailTemplate < CMS::Template
 
-  validates :system_name, presence: true
   validates :current, presence: true
   validates :system_name, uniqueness: { scope: %i[provider_id], allow_blank: true }
```
I reviewed this and found an error: `SystemName` sets uniqueness to `provider_id` and `type`, but not all models including a system name have a `type` column, (e.g. `File`). In fact, uniqueness by provider + type only makes sense for templates, so I moved that restriction to `CMS::Template` and set the uniqueness scope in `SystemName` to only `provider_id`. Thanks!

mayorova commented:
I like these changes, but we still need to update the existing DB objects, because some of them become invalid by these changes.

For example, many sections (including the built-in ones) have a space in the system name, and space is not allowed [here](https://github.com/3scale/porta/pull/3163/files#diff-86c790cda45e4a58997ce0998ef03a583854b2097e697a368b613278902a1ca6R9). So, if a user updates a section with a space in the system name, there will be a nasty error thrown.

Another breaking change (but not as critical, as at least the error will be visible to the users) is that when a user tries to update a template whose system name was previously empty, they will get a "can't be blank error", while it used to be OK (that's why it's breaking).

mayorova commented:
By the way, there is another bug I noticed, in the UI controller. Which is not related to this PR as such, but I just noticed it in relation to:

> Another breaking change (but not as critical, as at least the error will be visible to the users) is that when a user tries to update a template whose system name was previously empty, they will get a "can't be blank error", while it used to be OK (that's why it's breaking).

I'm putting it here so it's not lost... Will create a JIRA later.

When the page is saved for the first time, all form data go inside `cms_template`, e.g. 
```
cms_template[title]: page title
cms_template[path]: /page-path
cms_template[system_name]: 
...
cms_template[draft]: 
cms_template[published]: 
```

But when the update fails (e.g. because it fails validation, cause the `system_name` is blank), the conesequent requests (even with valid data) also fail, that's because the parameters now are wrapped in `cms_page` (all except `draft` and `published`), e.g. 

```
cms_page[title]: page title
cms_page[path]: /page-path
cms_page[system_name]: page-system-name
...
cms_template[draft]: 
cms_template[published]: 
```

This is weird and needs to be fixed (in another PR).

jlledom commented:
> For example, many sections (including the built-in ones) have a space in the system name, and space is not allowed [here](https://github.com/3scale/porta/pull/3163/files#diff-86c790cda45e4a58997ce0998ef03a583854b2097e697a368b613278902a1ca6R9). So, if a user updates a section with a space in the system name, there will be a nasty error thrown.

I changed the format to accept whitespaces, to avoid breaking changes. All new developer portals are created without whitespaces in their builtin templates, but older ones will still be valid too.

> Another breaking change (but not as critical, as at least the error will be visible to the users) is that when a user tries to update a template whose system name was previously empty, they will get a "can't be blank error", while it used to be OK (that's why it's breaking).

I updated the module so empty system names are also set on update. And wrote a test suite to test the whole logic.

jlledom commented:
> By the way, there is another bug I noticed, in the UI controller.

I created a Jira for it: https://issues.redhat.com/browse/THREESCALE-9316

mayorova commented:
> I updated the module so empty system names are also set on update. And wrote a test suite to test the whole logic.

Nice one @jlledom !

mayorova commented on app/models/cms/section.rb:
```diff
@@ -1,4 +1,5 @@
 class CMS::Section < ApplicationRecord
+  include CMS::SystemName
```
I'm afraid that adding the validation of `system_name` from the module will invalidate many sections existing in the DB, there are already too many in a "free form" :(

thalesmiguel commented on lib/tasks/cms/cms.rake:
```diff
@@ -210,6 +210,13 @@ namespace :cms do
+
+    desc 'Fixes invalid system names in sections'
+    task :section_invalid_system_names => :environment do
+      CMS::Section.unscoped.find_each do |section|
+        section.update_column(:system_name, "section-#{section.id}") unless section.valid?
```
By any chance, is there an easy way to figure out which sections are invalid with a DB Query so we don't have to go over all the records?
If not, I think the 1000 records batch size should be safe enough :v:

thalesmiguel replied:
If memory serves right, we can do something like `where("column ~* ?", 'regex')` BUT I'm not sure it works on MySQL (🤦) and having DB specific code should be avoided.
So this is cool for me!

thalesmiguel replied:
If memory serves right, we can do something like `where("column ~* ?", 'regex')` BUT I'm not sure it works on MySQL (🤦) and having DB specific code should be avoided.
So this is cool for me!

thalesmiguel commented on test/unit/tasks/cms/cms_test.rb:
```diff
@@ -22,4 +22,31 @@ def test_fix_pagination_href
+    test 'fixes invalid system names' do
+      section = FactoryBot.build(:cms_section, partial_path: '/', parent: @provider.sections.root, title: 'New section',
+                                 system_name: 'invalid name')
+      section.save!(validate: false)
```
If this doesn't have a chance to fail, we could use `save` instead:

```suggestion
      section.save(validate: false)
```

jlledom replied:
Commited

thalesmiguel commented on test/unit/tasks/cms/cms_test.rb:
```diff
@@ -22,4 +22,31 @@ def test_fix_pagination_href
+      assert_equal expected_sysname, section.reload.system_name
+    end
+
+    test "doesn't change valid system names" do
+      expected_sysname = "new-section"
```
We should use single quotes when there's no string interpolation.

jlledom replied:
Commited

thalesmiguel reviewed (approved):
So this new validation will invalidate some records on the DB but you also created the rake task to fix those up.
What's the deployment path for SaaS here? We first execute the rake task (the code) for SaaS and then we deploy this? Or maybe a migration?

jlledom commented:
> So this new validation will invalidate some records on the DB but you also created the rake task to fix those up. What's the deployment path for SaaS here? We first execute the rake task (the code) for SaaS and then we deploy this? Or maybe a migration?

You can't edit the records before deploying because the task calls to `valid?` which won't work properly unless the new code is deployed. Fortunately, the invalid records won't cause any problem until any operation is perfomed over them that needs validation (e.g. `#update`). So in my opinion we should deploy normally and then open a console in the pod and run:

 `RAILS_ENV=production bundle exec rake cms:fix:section_invalid_system_names`.

About running the task from a migration, not sure how to do it for OCP. In legacy servers, capistrano runs the migrations as part of the deploy process, but for OCP, AFAIK deploying consists on uploading a new container image and rolling out, I'm not sure when and how is the migration executed, so can't really say if a migration would be better.

akostadinov commented:
In such case, the condition can be changed to see if `system_name.blank?`. Why call full `valid if we only care about a single field?

mayorova commented:
I guess I'd vote for:
- Bringing back the whitespace to the format validator
- Either removing `CMS::SystemName` module from section model, or overriding the format to have something very generic that would not conflict the existing objects

So:
- going forward the `system_name` would be "uniform" and "nicely looking"
- we're not touching any existing objects and not invalidating them.

For on-premises upgrade the task would need to be run manually (or we need to transform it into a DB migration somehow), which is cumbersome.

I'd say it's too much work (not only for us, but for the customers) for little benefit.

jlledom commented:
> In such case, the condition can be changed to see if `system_name.blank?`. Why call full `valid if we only care about a single field?

That make sense.

> I guess I'd vote for:
> 
>     * Bringing back the whitespace to the format validator

No need to, thanks to the backfill task.

>     * Either removing `CMS::SystemName` module from section model, or overriding the format to have something very generic that would not conflict the existing objects
> 

I wouldn't remove the module inclusion from section, in that case there would be only one model including it, and the module existence wouldn't make much sense, the point of it is being reused from multiple models.

> 
> So:
> 
>     * going forward the `system_name` would be "uniform" and "nicely looking"

This wouldn't work if we remove `CMS::SystemName` from section.

> 
>     * we're not touching any existing objects and not invalidating them.
> 

So, remove the task?

> 
> For on-premises upgrade the task would need to be run manually (or we need to transform it into a DB migration somehow), which is cumbersome.
> 

The task can be converted to a DB migration easily. Doesn't have clients to run the task manually if we convert it to a migration?

> I'd say it's too much work (not only for us, but for the customers) for little benefit.

I vote for keeping the task, @akostadinov @thalesmiguel what do you say?

thalesmiguel commented:
> I vote for keeping the task

@jlledom So do I. It looks safe enough to be run as a migration, unless I'm missing some existing dark magic we have in place 🙈

jlledom commented:
Closing in favor of https://github.com/3scale/porta/pull/3248
